shell/packages/sandstorm-ui-powerbox/powerbox-server.js

```diff
@@ -1,16 +1,53 @@
+// Sandstorm - Personal Cloud Sandbox
+// Copyright (c) 2016 Sandstorm Development Group, Inc. and contributors
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+const Capnp = Npm.require("capnp");
+const Grain = Capnp.importSystem("sandstorm/grain.capnp");
+
 Meteor.methods({
-  newFrontendRef(sessionId, frontendRefVariety) {
+  newFrontendRef(sessionId, frontendRefVariety, saveLabel) {
     // Checks if the requester is an admin, and if so, provides a new frontendref of the desired
     // variety, provided by the requesting user, owned by the grain for this session.
     check(sessionId, String);
     check(frontendRefVariety, Match.OneOf(
-      { ipNetwork: Boolean },
-      { ipInterface: Boolean },
+      { ipNetwork: true },
+      { ipInterface: true },
     ));
+    check(saveLabel, {
+      defaultText: String,
+      localizations: Match.Optional([{ locale: String, text: String }]),
+    });
 
     const db = this.connection.sandstormDb;
-    if (!db.isAdmin(this.userId)) {
-      throw new Meteor.Error(403, "User must be an admin to powerbox offer frontendrefs");
+
+    let descriptor;
+    if (frontendRefVariety.ipNetwork) {
+      if (!db.isAdmin(this.userId)) {
+        throw new Meteor.Error(403, "User must be an admin to powerbox offer IpNetwork");
+      }
+
+      descriptor = "EAZQAQEAABEBF1EEAQH_QCAqemtXgqkAAAA";
```

Comment:
I would be happier if these were constructed programmatically to clarify intent and to make it easier for the next person who needs to produce such a descriptor to figure out how it should look.  Since this is server code, you have access to all the schemas and type IDs, which would make this clearer.